### PR TITLE
Source image tracking in tile generation

### PR DESCRIPTION
**Pull request by 20kdc**
This PR adds tracking to `tile.cpp` for if a source image is replaced at runtime.
If a source image is replaced, the generated images are replaced in turn.

This PR is Ready for Review.

# The below is inaccurate and is kept only for historical reasons.

This is in no way an absolute and complete solution, but I think it's a useful start and it reduces the likelihood of a memory leak caused by repeated uploads.

The problem this PR is intended to solve is that repeated uploads with differing paths are simultaneously necessary for large dynamically updating images and also cause a memory leak on receiving clients.

Note that this PR does *not* fully solve the existing issues with replacing textures at runtime. It does however begin that work by updating generated textures when one of their source images is replaced.

This leads to a number of situations where some form of "poke" is required to get a given target to update properly, but the results are usable.

This PR is Ready for Review.

Rationale on this:

There are various weird behaviours that will be seen here.
However to actually fix all of them would create a massive PR.

In practice I believe being able to safely runtime-replace textures with minor caveats is "good enough" for realistic usecases like player skins and so forth.

Still, I've kind of lost faith in the PR since beginning writing it, since there's a lot of edge cases that I don't think can be handled without changing stuff all over the place to the point where the PR ends up with a rather large scope.

I would not submit it at all, but I already made the fork and so forth, so I'm submitting it on the basis that engine developers will either agree with it despite it being incomplete or won't.

## How to test

1. Open up the devtest subgame.
2. Place down a `dynamic_media:node` node.
3. This is one of those "would be nice if this could change dynamically but won't happen" situations, it's not within scope for it to change in the following steps.
4. Use the entity spawner to create an `dynamic_media:entity` instance.
5. Right now this entity will update appearance to the current setting when it's reloaded.
6. Run commands: `/dynmedia 1` `/dynmedia 2`, `/dynmedia 1e`, `/dynmedia 2e`
7. Expected result is a HUD element on your screen appearing and changing image.
8. Create a second `dynamic_media:entity` instance
9. This will show the current state of the dynamic media image at when it's placed.
10. Run command `/dynmedia_ui`.
11. Expected result is a form - press the buttons to change the image sent to the client, press `reload` to send the client a slightly different formspec that causes it to reload the formspec and therefore update the image.

### CODE REVIEW

**Comment by appgurueu** [comment]
@Zughy I wouldn't call this maintenance - this is clearly a feature IMO.

**Comment by erlehmann** [comment]
Would this make it possiblle to dynamically update HUD textures without introducing memory leaks?
I have wanted to add explorable maps to mcl_maps and xmaps for a while, but so far did not do that.

**Comment by 20kdc** [comment]
> Would this make it possiblle to dynamically update HUD textures without introducing memory leaks? I have wanted to add explorable maps to mcl_maps and xmaps for a while, but so far did not do that.

That's the intent, yes. I don't have an absolute guarantee it prevents memory leaks in practice, but that falls under "fixable bugs" rather than the current situation.

**Comment by erlehmann** [comment]
@20kdc that sounds great. I fully agree with the intent of this change, especially if it degraded gracefully on older clients. Wolud it still spam older clients with unnecessary dynamic media though?

**Comment by 20kdc** [comment]
> @20kdc that sounds great. I fully agree with the intent of this change, especially if it degraded gracefully on older clients. Wolud it still spam older clients with unnecessary dynamic media though?

As I understand it it'll be in this situation where really old clients are spammed with dynamic media no matter what, somewhat old clients get dynamic media announcements and then will discard them, and recent clients will follow through as expected.

I imagine the ideal here is some sort of situation where maps are only sent to clients that need them.

It's still worth noting there's an alternative solution that presents itself where a packet or something is added to get the client to purge some media, but I didn't think of this until after writing this PR. It might be cleaner in regards to compatibility, but I'm not sure of the details. I think it's at least worth mentioning the idea because someone might decide it's a better approach than this PR.

**Comment by erlehmann** [comment]
> It's still worth noting there's an alternative solution that presents itself where a packet or something is added to get the client to purge some media

How would that approach deal with generated textures though?

**Comment by 20kdc** [comment]
> > It's still worth noting there's an alternative solution that presents itself where a packet or something is added to get the client to purge some media
> 
> How would that approach deal with generated textures though?

Presumably a specific packet to destroy a texture by generated texture ID.
(This is more flexible anyway because it allows purging embedded PNG textures.)

**Comment by appgurueu** [comment]
Relevant: https://github.com/minetest/minetest/issues/11640

Anyways, such a packet wouldn't quite suffice, since the sensible behavior is to regenerate all textures that depend on the replaced base texture. Thus, you'd have to keep an index `[base texture] = texmod` to regenerate texmods.

**Comment by sfan5** [comment]
My opinion: I don't think we should make this a supported feature yet because the "sometimes it works, sometimes it doesn't" isn't quite predictable but I looked through the code of this PR before (IIRC) and the groundwork can probably be merged for this to be made possible eventually in the future.

**Comment by 20kdc** [comment]
What parts should I strip out in order to make this mergable, if that'll be useful to engine devs?

**Comment by sfan5** [comment]
Basically just remove the API and code changes related to dynamic media and leave the rest (also cf. issue title).

**Comment by 20kdc** [comment]
I'm assuming that means "just the tile.cpp changes"

**Comment by 20kdc** [comment]
It is done.

**Review by sfan5 — APPROVED** [review]
Tested and works.

File: src/client/tile.cpp [hunk 1]
```diff
@@ -765,25 +807,36 @@ void TextureSource::rebuildImagesAndTextures()
...
 
-		video::IImage *img = generateImage(ti.name);
+void TextureSource::rebuildTexture(video::IVideoDriver *driver, TextureInfo &ti)
+{
+	if (ti.name.empty())
+		return; // this shouldn't happen, just a precaution
+
+	// replaces the previous sourceImages
+	// shouldn't really need to be done, but can't hurt
+	std::set<std::string> sourceImageNames;
```
Comment (SmallJoker) [inline src/client/tile.cpp:821]: 1. Variable names should use `snake_case` (reference https://dev.minetest.net/Code_style_guidelines#Miscellaneous )
2. Unless this is multithreaded ; wouldn't it be easier to move the `std::set` to a private `TextureSource` member variable? That avoids passing through the variable for each call.

Comment (20kdc) [reply src/client/tile.cpp:821]: Will correct 1.
2, while correct, complicates any attempt to multi-thread the code in future, and gives an incorrect idea of the nature of the variable. The variable is passed by reference (as it has to be) so no significant CPU load occurs.

**Review by SmallJoker — APPROVED** [review]
Looks good. Ran a quick test and found nothing unusual.

**Comment by 20kdc** [comment]
_The following is informational._

Just an FYI to anyone from the future wondering about the changes in this merge:

If the changes made here look a bit weird, it has to do with the history above of this PR.

It started out as a "buggy but useful" approach to the problem of how to send dynamic updatable textures without flooding the media cache (at least too much - `m_texture_trash` limits what can be done in this respect).

adf95601f18e64d8e9fd25b741347f3798c542e1 was an attempt to strip it down to just the texture tracking, but the `clientpackethandler.cpp` change to allow remote resource updates was still there, leaving `m_media_pushed_files` as an unused variable.

(It's worth noting that a patched server was still able to convince the client to double-load files even before the changes in this PR. The difference is that the original client would only accept this occurring once per filename. Of course, there are effectively infinite filenames, but a hypothetical future client still can't detect when a filename becomes unused, because the server might then send an entity or formspec that needs it.)

sfan5 then completed the removal of `m_media_pushed_files` and fixed some changes I hadn't properly cleaned up.

As such, any future server changes to allow remote texture updates as per the original version of this PR will presumably work on any client this merge and later, but there is a caveat.

Basically, if someone (maybe me in future, but probably not?) gets rid of `m_texture_trash` (mind that this requires some sort of `TextureRef` class used at all long-term texture use sites i.e. active objects - and some sort of notification mechanism due to Irrlicht's stateful nature), then real-time non-buggy texture updates are possible, along with collection of unused modifiers.
